### ISSUE TITLE
New logs stop being print out on console in Kubernetes in v4.0.5

### ISSUE THREAD

**Issue by AylaAsia-GaryJin** [emqx/emqx#3559]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
Environment
* OS: Kubenetes1.15.9
* Erlang/OTP: 22.1
* EMQ: 4.0.5
Description
A description of the issue
This is an old issue I raised in ekka section (emqx/ekka#2185). Since I recently upgraded EMQ to 3.0 and found out the issue still exists, I feel like it would be better to mention it here as well.
My EMQ server is hosted in Kubenetes. It is all good when I use it during the day. But after leaving it over the weekend, the new logs stop being print out on console. Although I can still see the new logs from erlang.log file.

Note:
* EMQ is builded from the source code and add an auth plugin based on the plugin-templete.
* EMQ cluster has 3 instances and all instances have this problem.

There're Dockerfile, docker-entrypoint.sh and start.sh:

### 1. Dockerfile
```
ARG BUILD_FROM=xxxxx.com.cn/buildbase/erlang-build:22.1-emqx
ARG RUN_FROM=xxxxx.com.cn/library/alpine:3.10
FROM ${BUILD_FROM} AS builder

ARG EMQX_VSN=v4.0.5
ARG EMQX_NAME=emqx
ARG DEPS_GITLAB_HOST=gitlab.ayla.com.cn
ARG DEPS_GITLAB_USER=gitlab-ci
ARG DEPS_GITLAB_TOKEN

WORKDIR /codes
COPY . .

ENV EMQX_DEPS_DEFAULT_VSN="$EMQX_VSN"

RUN make ${EMQX_NAME}

FROM $RUN_FROM

# Basic build-time metadata as defined at http://label-schema.org
LABEL org.label-schema.docker.dockerfile="Dockerfile" \
    org.label-schema.license="GNU" \
    org.label-schema.name="emqx" \
    org.label-schema.version=${EMQX_VSN} \
    org.label-schema.description="EMQ (Erlang MQTT Broker) is a distributed, massively scalable, highly extensible MQTT messaging broker written in Erlang/OTP." \
    org.label-schema.url="http://emqx.io" \
    org.label-schema.vcs-type="Git" \
    org.label-schema.vcs-url="https://github.com/emqx/emqx-docker" \
    maintainer="Raymond M Mouthaan <raymondmmouthaan@gmail.com>, Huang Rui <vowstar@gmail.com>, EMQ X Team <support@emqx.io>"

COPY --from=builder /codes/deploy/docker/docker-entrypoint.sh /usr/bin/
COPY --from=builder /codes/deploy/docker/start.sh /usr/bin/
COPY --from=builder /codes/_build/emqx*/rel/emqx /opt/emqx

RUN ln -s /opt/emqx/bin/* /usr/local/bin/ 
RUN apk add --no-cache curl ncurses-libs openssl sudo libstdc++

WORKDIR /opt/emqx

RUN adduser -D -u 1000 emqx \
    && echo "emqx ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers

RUN chgrp -Rf emqx /opt/emqx && chmod -Rf g+w /opt/emqx \
      && chown -Rf emqx /opt/emqx

USER emqx

VOLUME ["/opt/emqx/log", "/opt/emqx/data", "/opt/emqx/lib", "/opt/emqx/etc"]

# emqx will occupy these port:
# - 1883 port for MQTT
# - 8080 for mgmt API
# - 8083 for WebSocket/HTTP
# - 8084 for WSS/HTTPS
# - 8883 port for MQTT(SSL)
# - 11883 port for internal MQTT/TCP
# - 18083 for dashboard
# - 4369 for port mapping
# - 5369 for gen_rpc port mapping
# - 6369 for distributed node
EXPOSE 1883 8080 8083 8084 8883 11883 18083 4369 5369 6369

ENTRYPOINT ["/usr/bin/docker-entrypoint.sh"]

CMD ["/usr/bin/start.sh"]

```

### 2. docker-entrypoint.sh
```
#!/bin/sh
## EMQ docker image start script
# Huang Rui <vowstar@gmail.com>
# EMQ X Team <support@emqx.io>

## Shell setting
if [[ ! -z "$DEBUG" ]]; then
    set -ex
else
    set -e
fi

## Local IP address setting

LOCAL_IP=$(hostname -i |grep -E -oh '((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])'|head -n 1)

## EMQ Base settings and plugins setting
# Base settings in /opt/emqx/etc/emqx.conf
# Plugin settings in /opt/emqx/etc/plugins

_EMQX_HOME="/opt/emqx"

if [[ -z "$PLATFORM_ETC_DIR" ]]; then
    export PLATFORM_ETC_DIR="$_EMQX_HOME/etc"
fi

if [[ -z "$PLATFORM_LOG_DIR" ]]; then
    export PLATFORM_LOG_DIR="$_EMQX_HOME/log"
fi

if [[ -z "$EMQX_NAME" ]]; then
    export EMQX_NAME="$(hostname)"
fi

if [[ -z "$EMQX_HOST" ]]; then
    if [[ "$EMQX_CLUSTER__K8S__ADDRESS_TYPE" == "dns" ]] && [[ ! -z "$EMQX_CLUSTER__K8S__NAMESPACE" ]];then
        EMQX_CLUSTER__K8S__SUFFIX=${EMQX_CLUSTER__K8S__SUFFIX:-"pod.cluster.local"}
        DNSAddress="${LOCAL_IP//./-}.${EMQX_CLUSTER__K8S__NAMESPACE}.${EMQX_CLUSTER__K8S__SUFFIX}"
        export EMQX_HOST="$DNSAddress"
    elif [[ "$EMQX_CLUSTER__K8S__ADDRESS_TYPE" == "hostname" ]] && [[ ! -z "$EMQX_CLUSTER__K8S__NAMESPACE" ]]; then
        EMQX_CLUSTER__K8S__SUFFIX=${EMQX_CLUSTER__K8S__SUFFIX:-"svc.cluster.local"}
        HostAddress=$(sed -n "/^${LOCAL_IP}/"p /etc/hosts | grep -e "$(hostname).*.${EMQX_CLUSTER__K8S__NAMESPACE}.${EMQX_CLUSTER__K8S__SUFFIX}" -o)
        export EMQX_HOST="$HostAddress"
    else
        export EMQX_HOST="$LOCAL_IP"
    fi
fi

if [[ -z "$EMQX_WAIT_TIME" ]]; then
    export EMQX_WAIT_TIME=5
fi

if [[ -z "$EMQX_NODE__NAME" ]]; then
    export EMQX_NODE__NAME="$EMQX_NAME@$EMQX_HOST"
fi

if [[ -z "$EMQX_NODE_NAME" ]]; then
    export EMQX_NODE_NAME="$EMQX_NAME@$EMQX_HOST"
fi

# Set hosts to prevent cluster mode failed

# unset EMQX_NAME
# unset EMQX_HOST

if [[ -z "$EMQX_NODE__PROCESS_LIMIT" ]]; then
    export EMQX_NODE__PROCESS_LIMIT=2097152
fi

if [[ -z "$EMQX_NODE__MAX_PORTS" ]]; then
    export EMQX_NODE__MAX_PORTS=1048576
fi

if [[ -z "$EMQX_NODE__MAX_ETS_TABLES" ]]; then
    export EMQX_NODE__MAX_ETS_TABLES=2097152
fi

if [[ -z "$EMQX_LOG__TO" ]]; then
    export EMQX_LOG__TO="console"
fi

if [[ -z "$EMQX_LISTENER__TCP__EXTERNAL__ACCEPTORS" ]]; then
    export EMQX_LISTENER__TCP__EXTERNAL__ACCEPTORS=64
fi

if [[ -z "$EMQX_LISTENER__TCP__EXTERNAL__MAX_CONNECTIONS" ]]; then
    export EMQX_LISTENER__TCP__EXTERNAL__MAX_CONNECTIONS=1000000
fi

#if [[ -z "$EMQX_LISTENER__SSL__EXTERNAL__ACCEPTORS" ]]; then
#    export EMQX_LISTENER__SSL__EXTERNAL__ACCEPTORS=0
#fi
#
#if [[ -z "$EMQX_LISTENER__SSL__EXTERNAL__MAX_CONNECTIONS" ]]; then
#    export EMQX_LISTENER__SSL__EXTERNAL__MAX_CONNECTIONS=0
#fi

if [[ -z "$EMQX_LISTENER__WS__EXTERNAL__ACCEPTORS" ]]; then
    export EMQX_LISTENER__WS__EXTERNAL__ACCEPTORS=0
fi

if [[ -z "$EMQX_LISTENER__WS__EXTERNAL__MAX_CONNECTIONS" ]]; then
    export EMQX_LISTENER__WS__EXTERNAL__MAX_CONNECTIONS=0
fi


# Fix issue #42 - export env EMQX_DASHBOARD__DEFAULT_USER__PASSWORD to configure
# 'dashboard.default_user.password' in etc/plugins/emqx_dashboard.conf
if [[ ! -z "$EMQX_ADMIN_PASSWORD" ]]; then
    export EMQX_DASHBOARD__DEFAULT_USER__PASSWORD=$EMQX_ADMIN_PASSWORD
fi

# ayla specific ACL rules

if [[ -z "$EMQX_ACL_NOMATCH" ]]; then
    export EMQX_ACL_NOMATCH="deny"
fi

# Not allow the internal zone's clients to bypass authentication step
if [[ -z "$EMQX_ZONE__INTERNAL__BYPASS_AUTH_PLUGINS" ]]; then
    export EMQX_ZONE__INTERNAL__BYPASS_AUTH_PLUGINS=false
fi


# Catch all EMQX_ prefix environment variable and match it in configure file
CONFIG="${_EMQX_HOME}/etc/emqx.conf"
CONFIG_PLUGINS="${_EMQX_HOME}/etc/plugins"
for VAR in $(env)
do
    # Config normal keys such like node.name = emqx@127.0.0.1
    if [[ ! -z "$(echo $VAR | grep -E '^EMQX_')" ]]; then
        VAR_NAME=$(echo "$VAR" | sed -r "s/EMQX_([^=]*)=.*/\1/g" | tr '[:upper:]' '[:lower:]' | sed -r "s/__/\./g")
        VAR_FULL_NAME=$(echo "$VAR" | sed -r "s/([^=]*)=.*/\1/g")
        # Config in emq.conf
        if [[ ! -z "$(cat $CONFIG |grep -E "^(^|^#*|^#*\s*)$VAR_NAME")" ]]; then
            echo "$VAR_NAME=$(eval echo \$$VAR_FULL_NAME)"
            echo "$(sed -r "s/(^#*\s*)($VAR_NAME)\s*=\s*(.*)/\2 = $(eval echo \$$VAR_FULL_NAME|sed -e 's/\//\\\//g')/g" $CONFIG)" > $CONFIG   
        fi
        # Config in plugins/*
        for CONFIG_PLUGINS_FILE in $(ls $CONFIG_PLUGINS); do
            if [[ ! -z "$(cat $CONFIG_PLUGINS/$CONFIG_PLUGINS_FILE |grep -E "^(^|^#*|^#*\s*)$VAR_NAME")" ]]; then
                echo "$VAR_NAME=$(eval echo \$$VAR_FULL_NAME)"
                echo "$(sed -r "s/(^#*\s*)($VAR_NAME)\s*=\s*(.*)/\2 = $(eval echo \$$VAR_FULL_NAME|sed -e 's/\//\\\//g')/g" $CONFIG_PLUGINS/$CONFIG_PLUGINS_FILE)" > $CONFIG_PLUGINS/$CONFIG_PLUGINS_FILE
            fi 
        done
    fi
    # Config template such like {{ platform_etc_dir }}
    if [[ ! -z "$(echo $VAR | grep -E '^PLATFORM_')" ]]; then
        VAR_NAME=$(echo "$VAR" | sed -r "s/([^=]*)=.*/\1/g"| tr '[:upper:]' '[:lower:]')
        VAR_FULL_NAME=$(echo "$VAR" | sed -r "s/([^=]*)=.*/\1/g")
        echo "$(sed -r "s@\{\{\s*$VAR_NAME\s*\}\}@$(eval echo \$$VAR_FULL_NAME|sed -e 's/\//\\\//g')@g" $CONFIG)" > $CONFIG
    fi
done

## EMQ Plugin load settings
# Plugins loaded by default

if [[ ! -z "$EMQX_LOADED_PLUGINS" ]]; then
    echo "EMQX_LOADED_PLUGINS=$EMQX_LOADED_PLUGINS"
    # First, remove special char at header
    # Next, replace special char to ".\n" to fit emq loaded_plugins format
    echo $(echo "$EMQX_LOADED_PLUGINS."|sed -e "s/^[^A-Za-z0-9_]\{1,\}//g"|sed -e "s/[^A-Za-z0-9_]\{1,\}/\. /g")|tr ' ' '\n' > /opt/emqx/data/loaded_plugins
fi

exec "$@"
```

### 3. start.sh
```
#!/bin/sh
set -e

emqx_exit(){
    # tail erlang.log.*
    erlang_log=$(echo $(ls -t /opt/emqx/log/erlang.log.*) | awk '{print $1}')
    num=$(sed -n -e '/LOGGING STARTED/=' ${erlang_log} | tail -1)
    tail -n +$((num-2)) ${erlang_log}

    echo "['$(date -u +"%Y-%m-%dT%H:%M:%SZ")']:emqx exit abnormally"
    exit 1
}

## EMQ Main script

# Start and run emqx, and when emqx crashed, this container will stop

/opt/emqx/bin/emqx start || emqx_exit

tail -f /opt/emqx/log/erlang.log.1 &

# Sleep 5 seconds to wait for the loaded plugins catch up.
sleep 5

echo "['$(date -u +"%Y-%m-%dT%H:%M:%SZ")']:emqx start"

# monitor emqx is running, or the docker must stop to let docker PaaS know
# warning: never use infinite loops such as `` while true; do sleep 1000; done`` here
#          you must let user know emqx crashed and stop this container,
#          and docker dispatching system can known and restart this container.
NEW_LOG_NUM=2
IDLE_TIME=0
MGMT_CONF='/opt/emqx/etc/plugins/emqx_management.conf'
MGMT_PORT=$(sed -n -r '/^management.listener.http[ \t]=[ \t].*$/p' $MGMT_CONF | sed -r 's/^management.listener.http = (.*)$/\1/g')
while [[ $IDLE_TIME -lt 5 ]]; do
    IDLE_TIME=$((IDLE_TIME+1))
    if curl http://localhost:${MGMT_PORT}/status >/dev/null 2>&1; then
        IDLE_TIME=0
        # Print the latest erlang.log
        if [[ -f /opt/emqx/log/erlang.log.${NEW_LOG_NUM} ]];then
            tail -f /opt/emqx/log/erlang.log.${NEW_LOG_NUM} &
            [[ ! -z $(ps -ef |grep "tail -f /opt/emqx/log/erlang.log" | grep -vE "grep|erlang.log.${NEW_LOG_NUM}" | awk '{print $1}') ]] && kill $(ps -ef |grep "tail -f /opt/emqx/log/erlang.log" | grep -vE "grep|erlang.log.${NEW_LOG_NUM}" | awk '{print $1}') 
            NEW_LOG_NUM=$((NEW_LOG_NUM+1))
        fi
    else
        echo "['$(date -u +"%Y-%m-%dT%H:%M:%SZ")']:emqx not running, waiting for recovery in $((25-IDLE_TIME*5)) seconds"
    fi
    sleep 5
done
# If running to here (the result 5 times not is running, thus in 25s emqx is not running), exit docker image
# Then the high level PaaS, e.g. docker swarm mode, will know and alert, rebanlance this service

emqx_exit
```

**Comment by zhanghongtong** [comment]
Hi, @AylaAsia-GaryJin 

You mentioned in the description:

> the new logs stop being print out on console. 

I want to confirm how you get the log in the console, because as far as I know, the `emqx console` command cannot be run in the background

**Comment by AylaAsia-GaryJin** [comment]
> Hi, @AylaAsia-GaryJin
> 
> You mentioned in the description:
> 
> > the new logs stop being print out on console.
> 
> I want to confirm how you get the log in the console, because as far as I know, the `emqx console` command cannot be run in the background

I mean that I use "kubectl logs -f the-emq-pod-name" can't print the new logs.

**Comment by AylaAsia-GaryJin** [comment]
When the emq cluster start, logs could be print out, and it stop to print out several days later when use the same command.

**Comment by zhanghongtong** [comment]
Hi, @AylaAsia-GaryJin . Happy to receive your reply
When the file size of `erlang.log.1` exceeds the value of the environment variable `RUN_ERL_LOG_MAXSIZE`. `erlang.log.1` will not generate new content, but will generate `erlang.log.2`, in emqx, the default value of `RUN_ERL_LOG_MAXSIZE` is 1048576 bytes, I think this may be the reason why you cannot get new logs, for more information about the erlang.log file, you can view the [erlang documentation](http://erlang.org/doc/man/run_erl.html)

**Comment by AylaAsia-GaryJin** [comment]
> Hi, @AylaAsia-GaryJin . Happy to receive your reply
> When the file size of `erlang.log.1` exceeds the value of the environment variable `RUN_ERL_LOG_MAXSIZE`. `erlang.log.1` will not generate new content, but will generate `erlang.log.2`, in emqx, the default value of `RUN_ERL_LOG_MAXSIZE` is 1048576 bytes, I think this may be the reason why you cannot get new logs, for more information about the erlang.log file, you can view the [erlang documentation](http://erlang.org/doc/man/run_erl.html)

The log files are correct, but there is no new log print out through k8s log command

**Comment by zhanghongtong** [comment]
> The log files are correct, but there is no new log print out through k8s log command

This seems to be related to the image. You mentioned in the description that emqx is built by yourself, so can you share the `Dockerfile`, `docker-entrypoint.sh` and `start.sh` ?

**Comment by AylaAsia-GaryJin** [comment]
> > The log files are correct, but there is no new log print out through k8s log command
> 
> This seems to be related to the image. You mentioned in the description that emqx is built by yourself, so can you share the `Dockerfile`, `docker-entrypoint.sh` and `start.sh` ?

Update the three files in the description.

**Comment by zhanghongtong** [comment]
Hi, @AylaAsia-GaryJin, I see, this is a bug in `start.sh`, thank you for helping us find it

In `start.sh`, `NEW_LOG_NUM` is infinitely growing, but in fact, the number of `erlang.log` is controlled by `RUN_ERL_LOG_GENERATIONS`, the default is 5, so when `tail log/erlang.log.6`, `kubectl log` stops printing new log.

In order to verify my conjecture, please help me to observe the number of `log/erlang.log.*` in your container, and their update time, looking forward to your reply

**Comment by AylaAsia-GaryJin** [comment]
The log files of three nodes like this:
![image](https://user-images.githubusercontent.com/62411988/86088741-7c0b3400-bad9-11ea-955a-b61db2053942.png)

The config of log is default (size: 10M | file-num:5), and seems that file size limit is not working (only 1M).

**Comment by zhanghongtong** [comment]
Hi, @AylaAsia-GaryJin . Glad to get your reply

Like I said, the log problem is caused by the infinite growth of `NEW_LOG_NUM`, we will fix it in the next version, you can also modify `start.sh` to fix it by yourself.

**Comment by AylaAsia-GaryJin** [comment]
Hi, @zhanghongtong .any suggestions about how to modify the `start.sh`, because I don't well about shell, thank you.

**Comment by zhanghongtong** [comment]
> The config of log is default (size: 10M | file-num:5), and seems that file size limit is not working (only 1M).

The file size is only 1M is correct, because in [`bin/emqx`](https://github.com/emqx/emqx-rel/blob/413ce1409cde8e9142b6f7571097ae4ce4a6f268/bin/emqx#L321), `RUN_ERL_LOG_GENERATIONS=1048576` is set, we will modify `bin/emqx` in the next version, if the user sets `RUN_ERL_LOG_GENERATIONS`, then the user setting should prevail

**Comment by zhanghongtong** [comment]
> Hi, @zhanghongtong .any suggestions about how to modify the `start.sh`? Because I don't know well about shell, thank you.

I thought about it, and I didn't think of any particularly good way. Maybe set `log.rotation = off` in `etc/emqx.conf`, and `tail -f emqx.log.1` is a good way, I will give the code later

**Comment by zhanghongtong** [comment]
@AylaAsia-GaryJin Please check [this](https://github.com/emqx/emqx-rel/pull/539/files)

**Comment by AylaAsia-GaryJin** [comment]
> @AylaAsia-GaryJin Please check this
Thank you!

**Comment by zhanghongtong** [comment]
@AylaAsia-GaryJin You're welcome :)

**Comment by HJianBo** [comment]
Hi guys, it has included in v4.1.1

**Comment by HJianBo** [comment]
Sorry @AylaAsia-GaryJin perhaps I made a mistake. It just merged to 4.2-x, not v4.1.1.

When it merged to a new official version, I will notify you on this

**Comment by zhanghongtong** [comment]
HI, @AylaAsia-GaryJin 
It has included in v4.1.2

**Comment by AylaAsia-GaryJin** [comment]
> HI, @AylaAsia-GaryJin
> It has included in v4.1.2

Thank you.
